lib/osutil/osutil.go

```diff
@@ -26,12 +26,54 @@ var renameLock = sync.NewMutex()
 // TryRename renames a file, leaving source file intact in case of failure.
 // Tries hard to succeed on various systems by temporarily tweaking directory
 // permissions and removing the destination file when necessary.
-func TryRename(filesystem fs.Filesystem, from, to string) error {
+func TryRename(from fs.Filesystem, fromPath string, to fs.Filesystem, toPath string) error {
 	renameLock.Lock()
 	defer renameLock.Unlock()
 
-	return withPreparedTarget(filesystem, from, to, func() error {
-		return filesystem.Rename(from, to)
+	return withPreparedTarget(from, fromPath, to, toPath, func() error {
+		// Optimization
+		if from == to || to == nil || (from.Type() == to.Type() && from.URI() == to.URI()) {
+			return from.Rename(fromPath, toPath)
+		}
+
+		// See if one is a prefix of the other filesystem.
+		if from.Type() == to.Type() {
+			fromUri := from.URI()
+			toUri := to.URI()
+			if strings.HasPrefix(toUri, fromUri) {
+				newToPath := filepath.Join(strings.TrimPrefix(toUri, fromUri), toPath)
+				if err := from.Rename(fromPath, newToPath); err == nil {
+					return err
+				}
+			} else if strings.HasPrefix(fromPath, toPath) {
+				newFromPath := filepath.Join(strings.TrimPrefix(fromUri, toUri), fromPath)
+				if err := to.Rename(newFromPath, toPath); err == nil {
+					return err
+				}
+			} else {
```

Comment:
I don't think we should spend energy on cross filesystem renames. In general we shouldn't do them. There is a special case where this may be useful in some legacy cases like the staggered versioner when both filesystems are "basic", ad we can have a `RenameBetweenBasicFSs` that does the same as `os.Rename` - fails if the two `Filesystems` aren't in fact backed by the same actual underlying filesystem.

I don't want to think too much about what it would mean to "rename" a file from S3 into Upspin or whatever. There is no relevant sense in which it is still the same object with just a different name. The reasonable operation is copy followed by delete, with separate error handling on each.